### ISSUE TITLE
Add Complete and TryComplete methods to the System.Threading.Channels.ChannelReader

### ISSUE THREAD

**Issue by dmitriyse** [dotnet/corefx#29823]
The problem:
When consuming processing block has been stopped(due some error for example) and does not pulling ChannelReader the producing block does not know about that and still produces items and writes them into the ChannelWriter.

Producer ---> ChannelWriter==>Channel==>ChannelReader--X-->Consumer

```CSharp
public Task DoReading(ChannelReader<MyItem> reader)
{
    try
    {
        while(true)
        {
            // Do processing
        }
    }
    catch(Exception ex)
    {
        // Some how notify producer to stop, may be due to error.
    }
}
```

The solution:
Add symmetric methods like ChannelWriter.Complete and ChannelWriter.TryComplete
```CSharp
public class ChannelReader<T>
{
    // ...
    public abstract bool TryComplete(Exception error = null);
    public abstract void Complete(Exception error =null);
    // ...
}

public class ChannelWriter<T>
{
     // Completes when ChannelReader.Complete or ChannelReader.TryComplete is called.
     public abstract Task Completed { get; }
}

Items that are stored in the channel should be able to read even after Complete or TryComplete has been called.

With that feature large processing graphs can be organized many times simpler.

**Comment by davidfowl** [comment]
Yea, I brought this up in early prototypes and it does make sense to add it (if it can be added in a non breaking way, the way you propose is breaking). FWIW, System.IO.Pipelines.PipeReader and PipeWriter has the exact same API for this reason.

cc @stephentoub

**Comment by stephentoub** [comment]
I understand the use case, but I'm still not convinced it's desirable.  This effectively adds a channel in the opposite direction, at which point, why not just have a channel for the opposite direction?  I realize it's not quite that, but it seems like a slippery slope towards that, and I really don't want to see these things become bidirectional in that manner.

**Comment by davidfowl** [comment]
We should write that code sample, it's not very nice..

**Comment by AArnott** [comment]
I'm not sure I understand your concern, @stephentoub. Given you have `PipeWriter`/`PipeReader` as a precedent, it seems hard that it can be that bad. You already have the reader communicating with the writer for channels in terms of throttling. So this is just an extreme case of throttling where you're saying "there won't *ever* be any reading done again, so you can just forget about adding elements to the channel".

I just had a program hang because of this, where all the writers were waiting forever while my program was waiting for the completion of a dataflow block. I would prefer: the exception reported from the reader to cause all the `Channel<T>.Writer.WriteAsync` tasks to fault, leading to my dataflow blocks to fault, leading to my program to fail with the original exception instead of hang.

Without this feature, I'm going to have to create combined CancellationTokens for every single call into `WriteChannel<T>.WriteAsync` (which will incur allocations for the new CTS and its array of two cancellation tokens) so that when the reader fails I can manually cancel all those readers.

**Comment by stephentoub** [comment]
I'd like to see it prototyped to demonstrate that doing it correctly doesn't negatively impact throughput, in particular for channels like Channel.CreateUnbounded, where reading is lock-free as long as there's an item available.  If it can be done in an entirely opt-in / pay-for-play manner, where there's no negative throughput impact when the feature isn't used, it could be considered.  Of course the members couldn't be abstract, they'd need to be virtual, so please also define what the default behavior would be when not overridden.

> I just had a program hang because of this, where all the writers were waiting forever while my program was waiting for the completion of a dataflow block. I would prefer: the exception reported from the reader to cause all the Channel<T>.Writer.WriteAsync tasks to fault, leading to my dataflow blocks to fault, leading to my program to fail with the original exception instead of hang.

In your specific example, why can't your reader become a writer for the purposes of failing the block?

**Comment by AArnott** [comment]
> why can't your reader become a writer for the purposes of failing the block?

Interesting. I had dismissed that idea earlier because I assumed completion of the channel writer would only be considered after what was already in the queue had be flushed (since successful completion has that effect). If calling `ChannelWriter<T>.Complete(Exception)` will fault the channel immediately, leading to pending `WriteAsync` calls to throw with the exception, then that's a perfectly acceptable solution.

**Comment by stephentoub** [comment]
> If calling ChannelWriter<T>.Complete(Exception) will fault the channel immediately, leading to pending WriteAsync calls to throw with the exception, then that's a perfectly acceptable solution.

Yes.  Of all the built-in channels, the only one for which this is relevant is the bounded channel (since there will never be a pending write on an unbounded channel, since it'll always complete synchronously), and that'll wake up any pending writes when the channel is completed:
https://github.com/dotnet/corefx/blob/a51d39a69de02682c53748c35473803252edbd2c/src/System.Threading.Channels/src/System/Threading/Channels/BoundedChannel.cs#L282-L290

**Comment by AArnott** [comment]
Confirmed: that resolves my issue. Thanks.
